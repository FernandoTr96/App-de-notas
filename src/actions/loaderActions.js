import { types } from "../types/types"

export const startLoader = ()=>{
    return {
        type: types.loaderTrue,
        payload:{
            value: true
        }
    }
}

export const finishLoader = ()=>{
    return {
        type: types.loaderFalse,
        payload:{
            value: false
        }
    }
}

export const purgeLoader = ()=>(
    {
        type: types.purgeLoader
    }
)