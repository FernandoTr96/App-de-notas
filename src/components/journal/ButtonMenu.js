import React from 'react'

export const ButtonMenu = () => {
    return (

        <label htmlFor='btnMenu' className='journal__btn-menu'>
            <span className='line-1'></span>
            <span className='line-2'></span>
            <span className='line-3'></span>
        </label>

    )
}
