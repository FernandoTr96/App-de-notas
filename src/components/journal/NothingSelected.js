import React from 'react'

export const NothingSelected = () => {
    return (
        <div className='nothing__main-content'>
            <p>
                select something
                <br/>
                or
                <br/>
                create entry
                <br/>
                <br/>
                <img 
                width='50'
                height='50'
                src='https://c.tenor.com/0jWPxOP-71UAAAAC/chibi-cat.gif' alt='gif' />
            </p>
        </div>
    )
}
