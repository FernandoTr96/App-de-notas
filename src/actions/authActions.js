import { 
    auth,
    signInWithPopup,
    googleAuthProvider,
    createUserWithEmailAndPassword,
    updateProfile,
    signInWithEmailAndPassword,
    signOut
} from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoader, startLoader } from "./loaderActions";
import Swal from 'sweetalert2';


export const startLoginEmailPassword = (email,password) =>{
    return async (dispatch)=>{

        await dispatch(startLoader());
        
        await signInWithEmailAndPassword(auth,email,password)
        .then((userCred) =>{

            const  {user:{uid,displayName,photoURL}} = userCred;
            dispatch(login(uid,displayName,photoURL));

        })
        .catch(err =>{
            console.log(err);
            Swal.fire({
                title: 'Error!',
                text: err,
                icon: 'error',
                confirmButtonText: 'ok'
            })
        })

        await dispatch(finishLoader());
    }
};

export const startGoogleLogin = () =>{
    return (dispatch)=>{
        signInWithPopup(auth,googleAuthProvider)
        .then( userCred =>{
            //console.log(userCred);
            const  {user:{uid,displayName,photoURL}} = userCred;
            dispatch(login(uid,displayName,photoURL));
        })
    }
};

export const login = (uid,displayName,photoURL='https://placeimg.com/100/100/animals') => (
    {
        type: types.login,
        payload:{
            uid,
            displayName,
            photoURL
        }
    }
);

export const registerUser = (email,password,userName,reset)=>{
    return async (dispatch)=>{
        
        await dispatch(startLoader());

        //crear usuario en firebase
        await createUserWithEmailAndPassword(auth,email,password)
        .then( async ({user}) => {
            
            //funcion defirebase al obtener el usuario
            //permite registrar el nombre y la foto
            await updateProfile(auth.currentUser,{
                displayName: userName,
                photoURL: 'https://placeimg.com/100/100/any'
            });

            //actualizar el estado global
            const  {uid,displayName,photoURL} = user;
            dispatch(login(uid,displayName,photoURL));

            //reset form
            reset();

        })
        .catch(err =>{
            Swal.fire({
                title: 'Error!',
                text: err,
                icon: 'error',
                confirmButtonText: 'ok'
            })
        })

        await dispatch(finishLoader());
        
    }
};

export const startLogout = ()=>{
    return async (dispatch) =>{
        await signOut(auth);
        dispatch(logout());
    }
}

export const logout = ()=>(
    {
        type: types.logout
    }
)
