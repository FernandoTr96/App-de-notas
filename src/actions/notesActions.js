import Swal from "sweetalert2";
import { db, addDoc, collection, getDocs, updateDoc, doc, deleteDoc, query, orderBy } from "../firebase/firebase-config";
import { uploadImg } from "../helpers/uploadImg";
import {removeImg} from '../helpers/removeImg'
import { types } from "../types/types";
import { finishLoader, startLoader } from "./loaderActions";

//insertar en firestore addDoc

/* PARA INSERTAR EN FIRESTORE DEBE ESTAR AUTENTICADO    
    allow read, write: if request.auth != null;
*/

export const startNewNote = () => {
    //getState: permite traer el estado global como un useSelector()
    //estos argumentos son pasados por el moddleware
    return async (dispatch, getState) => {

        //id del usuario que llamo la accion o esta en sesion
        const uid = getState().auth.uid;

        //objeto que contendra la info de la nota
        const newNote = {
            title: 'New note',
            body: '...',
            date: new Date().getTime()
        }

        //insertar en firestore
        try {

            dispatch(startLoader());

            const docRef = await addDoc(
                collection(db, `${uid}/journal/notes`),
                newNote
            );

            //activar nota
            dispatch(activeNote(docRef.id, newNote));

            //obtener las notas para actualizar el estado
            dispatch(getNotes(uid));

            dispatch(finishLoader());


        } catch (error) {

            console.log(error);
            Swal.fire({
                title: 'Not created',
                text: error,
                icon: 'error',
                confirmButtonText: 'ok'
            })

        }

    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const resetActiveNote = () => ({
    type: types.notesResetActive
})


//Leer datos de firestore getDocs

export const getNotes = (uid) => {
    return async (dispatch) => {

        let notes = [];

        try {

            const q = query(collection(db, `${uid}/journal/notes`), orderBy('date', 'desc'));
            const data = await getDocs(q);

            data.forEach((doc) => {
                notes.push(
                    {
                        id: doc.id,
                        ...doc.data()
                    }
                )
            })

            dispatch(setNotes(notes));

        } catch (error) {

            Swal.fire({
                title: 'I cant get the notes',
                text: error,
                icon: 'error',
                confirmButtonText: 'ok'
            })

        }

    }
}

export const setNotes = (notes) => (
    {
        type: types.notesSet,
        payload: notes
    }
);


//actualizar nota
export const updateNote = (note) => {
    return async (dispatch, getState) => {

        const { auth: { uid } } = getState();
        const noteToFireStore = { ...note };
        delete noteToFireStore.id;

        try {

            const docRef = doc(db, `${uid}/journal/notes`, note.id);
            await updateDoc(docRef, noteToFireStore);
            dispatch(activeNote(note.id, note));
            dispatch(refreshNote(note.id, note));

            Swal.fire({
                title: 'Note updated',
                text: 'The note has been updated successfully !!',
                icon: 'success',
                confirmButtonText: 'ok'
            })

        } catch (error) {

            Swal.fire({
                title: 'I cant update the note',
                text: error,
                icon: 'error',
                confirmButtonText: 'ok'
            })

        }
    }
}

const refreshNote = (id, data) => ({
    type: types.noteRefresh,
    payload: {
        id, data
    }
})


//subir imagen
export const uploadImage = (file) => {
    return async (dispatch, getState) => {

        dispatch(startLoader());

        const upload = await uploadImg(file);

        if (upload !== null) {

            const {url, public_id} = upload;

            const noteActive = getState().notes.active;
            const noteWithUrlToFirestore = {
                ...noteActive,
                public_id,
                url
            };

            if (getState().notes.active?.public_id) {
                const public_id = getState().notes.active.public_id;
                removeImg(public_id);
            }

            dispatch(activeNote(noteWithUrlToFirestore.id, noteWithUrlToFirestore));
            dispatch(updateNote(noteWithUrlToFirestore));

        }

        dispatch(finishLoader());

    }
}

//Eliminar nota

export const deleteNote = (id) => {
    return (dispatch, getState) => {

        Swal.fire({
            title: 'Are you sure?',
            text: "You won't be able to revert this!",
            icon: 'warning',
            showCancelButton: true,
            confirmButtonColor: '#3085d6',
            cancelButtonColor: '#d33',
            confirmButtonText: 'Yes, delete it!'
        }).then(async (result) => {
            if (result.isConfirmed) {

                dispatch(startLoader());

                try {

                    const uid = getState().auth.uid;
                    await deleteDoc(doc(db, `${uid}/journal/notes`, id));
                    dispatch(removeNote(id));

                    if (getState().notes.active?.public_id) {
                        const public_id = getState().notes.active.public_id;
                        removeImg(public_id);
                    }

                    Swal.fire(
                        'Deleted!',
                        'Your note has been deleted.',
                        'success'
                    )


                } catch (error) {

                    Swal.fire({
                        title: 'i cant delete the note',
                        text: error,
                        icon: 'error',
                        confirmButtonText: 'ok'
                    })

                }

                dispatch(finishLoader());

                dispatch(resetActiveNote());

            }
        })
    }
}

const removeNote = (id) => (
    {
        type: types.noteRemove,
        payload: id
    }
)

export const purgeNotes = () => (
    {
        type: types.purgeNotes
    }
)

/* ASI ENTIENDO NO SQL

//coleccion
uid = [
    //documento
    journal = {

        //coleccion
        notes = [

            //documento
            id = {

                //campos
                body: "",
                title: "",
                date: ""
            }
        ]
    }
]

[
    {
        journal: {
            notes:[
                01:{
                    body: "",
                    title: "",
                    date: ""
                }
            ]
        }
    }
]*/