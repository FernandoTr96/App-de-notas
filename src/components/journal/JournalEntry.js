import React from 'react'
import {useDispatch} from 'react-redux'
import { activeNote } from '../../actions/notesActions';


export const JournalEntry = ({id,title,body,date,url,checkboxMenu}) => {

   const dispatch = useDispatch();
   const dateFormat = new Date(date).toLocaleDateString();

    const handleActiveNote = ()=>{
        
        dispatch(activeNote(id,{
           title,
           body,
           date,
           url
        }));

        //cerrar el menu usando la referencia si esta en movil
        checkboxMenu.current.checked = false;
    }


    return (
        <div onClick={handleActiveNote} className='journal__entry animate__animated animate__bounceInDown'>           
            {
                url ?
                <div className='journal__picture' style={{backgroundImage: `url(${url})` }}></div> :
                <div className='journal__picture' style={{backgroundImage:'url(https://socialistmodernism.com/wp-content/uploads/2017/07/placeholder-image.png?w=640)'}}></div>
            }
            <div className='journal__body'>
                <div className='journal__title'>
                    <h4>{title}</h4>
                </div>
                <div className='journal__desc'>
                    <small>{body}</small>
                </div>
            </div>
            <div className='journal__date'>
                <small>{ dateFormat }</small>
            </div>
        </div>
    )
}
