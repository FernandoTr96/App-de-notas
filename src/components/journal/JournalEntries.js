import React from 'react'
import { JournalEntry } from './JournalEntry';
import { useSelector } from 'react-redux'

export const JournalEntries = ({checkboxMenu}) => {

    const {notes:{notes}} = useSelector(state => state);

    return (
        <div className='journal__entries'>
            {
                notes.map(note => 
                    <JournalEntry 
                    key={note.id} 
                    {...note}
                    checkboxMenu={checkboxMenu}
                    />
                )
            }
        </div>
    )
}
