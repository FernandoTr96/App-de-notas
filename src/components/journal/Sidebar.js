import React, { useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startLogout } from '../../actions/authActions'
import { purgeNotes, startNewNote } from '../../actions/notesActions'
import { ButtonMenu } from './ButtonMenu'
import { JournalEntries } from './JournalEntries'
import { Loader } from '../loading/Loader'
import { purgeLoader } from '../../actions/loaderActions'

export const Sidebar = () => {

    // useSelector de redux permite obtener el estado global
    // osea el conjunto de reducers
    const { auth: { displayName, photoURL }, loader: { loading } } = useSelector(state => state);

    const dispatch = useDispatch();

    const checkboxMenu = useRef();

    const handleLogout = () => {
        dispatch(startLogout());
        dispatch(purgeLoader());
        dispatch(purgeNotes());
    }

    const handleAddNewEntry = () => {
        dispatch(startNewNote());
    }


    return (
        <>
            <input ref={checkboxMenu} type='checkbox' id='btnMenu' />
            <ButtonMenu />
            <aside className='journal__aside animate__animated animate__fadeIn'>
                <div className='journal__user'>
                    <div className='journal__displayName'>
                        <img className='journal__photo' src={photoURL} alt='photoURL' />
                        <small>
                            {displayName}
                        </small>
                    </div>
                    <button onClick={handleLogout} ><i className="fas fa-power-off"></i> Logout</button>
                </div>

                <div className='journal__btn-add-entry' title='add entry'>
                    {

                        loading ?
                        <Loader src='https://samherbert.net/svg-loaders/svg-loaders/tail-spin.svg' w={25} h={25} /> :
                        <button onClick={handleAddNewEntry}>
                            <i className="far fa-calendar-plus"></i>
                            <small>Create entry</small>
                        </button>

                    }
                </div>

                <JournalEntries checkboxMenu={checkboxMenu} />
            </aside>
        </>
    )
}
