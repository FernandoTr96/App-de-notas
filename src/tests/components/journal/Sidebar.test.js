import React from 'react';
import {mount} from 'enzyme'
import {MemoryRouter} from 'react-router-dom'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'       
import { Provider } from 'react-redux';
import { Sidebar } from '../../../components/journal/Sidebar';
import { purgeNotes, startNewNote } from '../../../actions/notesActions';
import { startLogout } from '../../../actions/authActions';
import { purgeLoader } from '../../../actions/loaderActions';

const middlewares = [thunk];
const mockstore = configureStore(middlewares);
const init = {
    auth:{
        displayName: 'Fernando',
        photoURL: 'https://placeimg.com/100/100/any'
    },
    loader:{
        loading: false
    },
    notes:{
        notes: []
    }
};
const store = mockstore(init);
store.dispatch= jest.fn();

let wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <Sidebar/>
        </MemoryRouter>
    </Provider>
);


//mocks
jest.mock('../../../actions/notesActions', ()=>({
    startNewNote: jest.fn(),
    purgeNotes: jest.fn()
}))

jest.mock('../../../actions/loaderActions', ()=>({
    purgeLoader: jest.fn()
}))

jest.mock('../../../actions/authActions.js', ()=>({
    startLogout: jest.fn()
}))


describe('Pruebas en <Sidebar/>', () => {
    
    test('debe renderizar correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe mostrar el nombre de ususario ', () => {   
        const displayNameSnap = wrapper.find('.journal__displayName small').text().trim();
        expect(displayNameSnap).toBe(init.auth.displayName);
    });

    test('addNewNote debe lanzar la accion de una nueva nota ', () => {
        
        wrapper.find('.journal__btn-add-entry button').simulate('click')
        expect(startNewNote).toHaveBeenCalled();

    });

    test('logout debe cerrar sesion y purgar los estados', () => {
        
        wrapper.find('.journal__user button').simulate('click');
        expect(startLogout).toHaveBeenCalled();
        expect(purgeLoader).toHaveBeenCalled();
        expect(purgeNotes).toHaveBeenCalled();

    });

});