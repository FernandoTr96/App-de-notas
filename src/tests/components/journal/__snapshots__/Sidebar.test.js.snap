// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Sidebar/> debe renderizar correctamente 1`] = `
Array [
  <input
    id="btnMenu"
    type="checkbox"
  />,
  <label
    className="journal__btn-menu"
    htmlFor="btnMenu"
  >
    <span
      className="line-1"
    />
    <span
      className="line-2"
    />
    <span
      className="line-3"
    />
  </label>,
  <aside
    className="journal__aside animate__animated animate__fadeIn"
  >
    <div
      className="journal__user"
    >
      <div
        className="journal__displayName"
      >
        <img
          alt="photoURL"
          className="journal__photo"
          src="https://placeimg.com/100/100/any"
        />
        <small>
          Fernando
        </small>
      </div>
      <button
        onClick={[Function]}
      >
        <i
          className="fas fa-power-off"
        />
         Logout
      </button>
    </div>
    <div
      className="journal__btn-add-entry"
      title="add entry"
    >
      <button
        onClick={[Function]}
      >
        <i
          className="far fa-calendar-plus"
        />
        <small>
          Create entry
        </small>
      </button>
    </div>
    <div
      className="journal__entries"
    />
  </aside>,
]
`;
